### DISCUSSION TITLE
Arbitrary value for bg-[url(...)] with data:image/svg-xml throws CSS syntax error

### DISCUSSION THREAD

**Discussion by thomasweitzel** [tailwindlabs/tailwindcss#12322] [Help] [answered]
I'm using Tailwind 3.3.5 and have this in my `tailwind.css` file, which works perfectly:

```
@tailwind base;
@tailwind components;
@tailwind utilities;

@layer base {
    blockquote::before {
        background-image: url('data:image/svg+xml;utf8,<svg viewBox="0 0 100 100" xmlns="http://www.w3.org/2000/svg"><circle cx="50" cy="50" r="40" stroke="red" fill="transparent" stroke-width="10"/></svg>');
        @apply absolute content-[''] w-[64px] left-[-84px] top-0 bottom-0 border-gray-600 bg-no-repeat;
    }
    blockquote {
        @apply relative ml-[84px] border-l-4 pl-4 border-gray-600 italic;
    }
}
```

When a blockquote is rendered, it looks like this:

![](https://pureandroid.com/downloads/files/misc/blockquote-css.png)

Now, I want to move the `background-image` to the `@apply` line and use the Tailwind CSS [arbitrary value feature](https://tailwindcss.com/docs/background-image#arbitrary-values). The line now reads as follows:

```
blockquote::before {
    @apply absolute content-[''] w-[64px] left-[-84px] top-0 bottom-0 border-gray-600 bg-no-repeat bg-[url('data:image/svg+xml;utf8,<svg viewBox="0 0 100 100" xmlns="http://www.w3.org/2000/svg"><circle cx="50" cy="50" r="40" stroke="red" fill="transparent" stroke-width="10"/></svg>')];
}
```

While I think this should work ok, it doesn't. It throws this error:

```
CssSyntaxError:
  tailwindcss: /tailwind/tailwind.css:1:1:
  The `bg-[url('data:image/svg+xml;utf8,<svg` class does not exist.
  If `bg-[url('data:image/svg+xml;utf8,<svg` is a custom class,
  make sure it is defined within a `@layer` directive.

  <stacktrace omitted>

  reason: "The `bg-[url('data:image/svg+xml;utf8,<svg` class does not exist.
  If `bg-[url('data:image/svg+xml;utf8,<svg` is a custom class,
  make sure it is defined within a `@layer` directive."
```

What I'm doing wrong? The last part (`defined within a `@layer` directive`) is definitely not the problem, right?

**Comment by wongjn** [comment, marked as answer]
Spaces delimit class names in HTML, and thus, they do as well in `@apply`. Thus, the class is `bg-[url('data:image/svg+xml;utf8,<svg` which - as the the error states - cannot be found. There is no benefit in using `@apply` for this, so consider using a regular CSS declaration:

```css
blockquote::before {
  @apply …;
  background-image: url(…);
}
```

As an aside, Adam Wathan, creator of Tailwind does advocate avoiding `@apply`:
https://twitter.com/adamwathan/status/1226511611592085504
https://twitter.com/adamwathan/status/1559250403547652097

**Reply by thomasweitzel** [reply]
Thanks for the hint with the spaces! Since the "regular" CSS declaration works like a charm, I think I have no choice.

Regarding `@apply`: yeah, I have seen it ;-). I'm styling my Hugo-powered blog, where the HTML is rendered from Markdown (.md) files. `@apply` is a convenient way to get the Tailwind CSS to the right places. Would you recommend another way of doing it? Is there one?

**Reply by wongjn** [reply]
I would recommend using regular CSS declarations where possible, using [`theme()` to reference Tailwind configuration values](https://tailwindcss.com/docs/functions-and-directives#theme), as per [Adam Wathan's recommendation](https://twitter.com/adamwathan/status/1559250403547652097).
